### ISSUE TITLE
files not published when added to the files property in package.json on Node.js v6

### ISSUE THREAD

**Issue by ashleygwilliams** [npm/npm#12498]
`index.js` if listed in the `"files"` attribute of `package.json` will be ignored on publish. files that are listed but NOT named `index.js` are not ignored.

test package: https://github.com/ashleygwilliams/index-node6-meta

## steps to reproduce

On Node v`6.0.0`, npm versions `3.8.6` and `2.15.4`, with OSX
1. fork, clone, `npm install` the `ashleygwilliams/index-node6-meta` repo

Note:
- package `index-node6@v1` does not list `index.js` in `"files"`, has a `index.js`
- package `index-node6@v2` does list `index.js` in `"files"`, does not have a `index.js`
- package `test-test-test2` does list `poop.js` in `"files"`, has a `poop.js` and `index.js`

## UPDATE:
After a period of time I cannot reproduce this. @___@

**Comment by ashleygwilliams** [comment]
likely related to this issue: https://github.com/npm/npm/issues/5082

**Comment by SamVerschueren** [comment]
It's not only with `index.js`. It seems all the files in the `files` property are not present in the published package. https://github.com/SamVerschueren/npm-publish-index-test/blob/v0.2.0/package.json

```
$ npm cache add git+https://github.com/SamVerschueren/npm-publish-index-test.git
$ npm cache ls npm-publish-index-test
~/.npm/npm-publish-index-test/npm-publish-index-test/
~/.npm/npm-publish-index-test/npm-publish-index-test/0.1.0/
~/.npm/npm-publish-index-test/npm-publish-index-test/0.1.0/package.tgz
~/.npm/npm-publish-index-test/npm-publish-index-test/0.1.0/package/
~/.npm/npm-publish-index-test/npm-publish-index-test/0.1.0/package/package.json
```

The files are not part of the tarball.

**Comment by SamVerschueren** [comment]
Not sure if the above commands actually do anything useful. I always get that output, even after I removed the `files` property. But I can confirm that removing the `files` property actually solves the bug.

**Comment by iarna** [comment]
@SamVerschueren That output is what I would expect. That's what the cache folder is supposed to look like.  You need to look inside `~/.npm/npm-publish-index-test/npm-publish-index-test/0.1.0/package.tgz
` to see if it's missing any of the files you need.

**Comment by SamVerschueren** [comment]
@iarna Oh ok, my bad. When extracting the `tgz` files I get the expected result. `0.1.0` and `0.2.0` does not have the js files while `0.3.0` does.

**Comment by SamVerschueren** [comment]
The title should say something like:

> files not published when added to the `files` property in `package.json` on Node.js v6

They aren't being installed because they aren't being published in the first place.

**Comment by ashleygwilliams** [comment]
updated!

**Comment by SamVerschueren** [comment]
I digged in the tar creation code but it's quite difficult to really understand what is going on there :). So it appears the `index.js` file is always missing if I add the files to the `files` property in `package.json`. But the file specified as `main` should always be packed into the tarball which is not happening either.

```json
"main": "index.js",
"files": [
    "lib.js"
]
```

`lib.js` is present in the tarball, `index.js` is not.

**Comment by iamstarkov** [comment]
@SamVerschueren i was thinking that `files` are only about whitelisting files to include into tarball, and `main` is about pointing to the module's entrypoint — pretty solid and explicit behaviour. So from my point of view i would not expect `index.js` to be published if its not in the `files` array.

**Comment by SamVerschueren** [comment]
I bumped upon this behaviour when reading the sourcecode.

https://github.com/npm/fstream-npm/blob/master/fstream-npm.js#L109-L111

**Comment by SamVerschueren** [comment]
Untested so not sure if it really behaves that way.

**Comment by othiym23** [comment]
@SamVerschueren I've spent a couple hours in that function, and its callers, and haven't yet been able to construct a scenario in which it does the wrong thing. (But, you know, I did spend a couple hours in there, so your intuition is sound!) `.applyIgnores()` is a synchronous function, so it would have to be a scenario in which its result gets ignored somehow. [On Twitter](https://twitter.com/evanhlucas/status/725710107015827457?lang=en), @evanlucas suspects that it may be a stitch getting dropped between two parts of npm's tar code, but it's still quite possibly an `fstream` problem, becase `node-tar`, which is built on `fstream` as well.

I'm looking at #5082 (which I currently believe this to be a duplicate of, but which has been exacerbated for some users by the release of Node 6.0.0 for as-yet undetermined reasons), and @kkoopa has been having problems there with `nan` that I'm trying to reproduce locally. My biggest problem is that I can't reliably trigger the bug myself (and never have been). This is why I was asking you and @sindresorhus what OS you run yesterday, because I too am on OS X and am trying to eliminate environmental differences that might be triggering this bug.

As I mentioned on Twitter and #5082, I have a [test repo](https://github.com/othiym23/eliminate-5082) that I'm using to automate testing (and which I'm running against an instrumented copy of npm) – if you could pull down the latest version of that code and run it and let me know if you get any failures, that would be very helpful. I discovered last night that two of the test packages had the same name (I copied them around and overlooked an update), which was triggering a far more boring race condition in the tests. Since fixing that, I haven't been able to get a failure at all.

**Comment by othiym23** [comment]
Also just noticed @SamVerschueren's [tweet](https://twitter.com/SamVerschueren/status/725620683435302912) about `node_modules` being required to reproduce. I'll incorporate that into my test scenarios. @SamVerschueren, are you bundling anything? There's a fair amount of special-purpose code related to bundling in that chunk of the code that it would be nice to eliminate from suspicion, if possible.

**Comment by SamVerschueren** [comment]
Did a PR in your test repo that fails on my system.

**Comment by kkoopa** [comment]
Using Node 6, I just tried installing and running eliminate-5082. All runs passed. I then immediately went back to packing NAN. Now the first attempt succeeded. However, no subsequent attempt ever succeeded.
I then went back to eliminate-5082, without reinstalling it. All runs passed. Going back to packing NAN, nothing succeeded, not even the first run. Tried cloning eliminate-5082 again and installing it anew, all tests passed. Went back to packing NAN, still no success.

In conclusion, I could not get eliminate-5082 to fail a single time, but my problematic package only succeeded once, the first time I tried it. It seems that packing only works once a day.

**Comment by SamVerschueren** [comment]
`eliminate-5082` currently does not take `.gitignore` and `node_modules` into account. When I add those as in this PR https://github.com/othiym23/eliminate-5082/pull/1, the `index-in-files` test files.

**Comment by othiym23** [comment]
@SamVerschueren I've merged your change and bingo! I'm seeing the same failure now. Excellent! Since I'm working against an instrumented version of npm (I'll throw up the patch later today), I have lots of interesting output to go through, but this should help quite a bit. Thank you!

**Comment by othiym23** [comment]
[Here's the output](https://gist.github.com/othiym23/7ecd556a61e76bfd68d60d3c8efbe059) of the tests when using my instrumented version of npm. You can clearly see in the `index-in-files` case that `index.js` isn't making it to the tar file (because I've got a [stream spy](https://www.npmjs.com/package/through2-spy) dumping the raw output to the console). Now to trace backwards and figure out where it's getting dropped.

**Comment by SamVerschueren** [comment]
Nice :)

**Comment by othiym23** [comment]
This is very much the same issue as #5082, so I'm closing it as a duplicate of that issue. With the work that @evanlucas and @addaleax have done today, we're getting very close to a real fix, but having the discussion split across two issues is starting to get a little confusing. Thanks to @ashleygwilliams for jumping on top of this, and thanks to all of you for the information and your patience!
